### PR TITLE
Improve service resilience and reduce timeout

### PR DESCRIPTION
**Pull request by anthony-britton-moj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppschallengesupportinterventionplanapi/client/RetryNetworkExceptions.kt [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package uk.gov.justice.digital.hmpps.hmppschallengesupportinterventionplanapi.client
+
+import org.springframework.web.reactive.function.client.WebClientResponseException
+import reactor.core.publisher.Mono
+import reactor.util.retry.Retry
+import java.io.IOException
+import java.time.Duration
+
+fun <T> Mono<T>.retryNetworkExceptions(): Mono<T> =
```
Comment (ReedSoftware) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppschallengesupportinterventionplanapi/client/RetryNetworkExceptions.kt:9]: Are we confident that every endpoint we have and are going to have is idempotent, if we're retrying on things like 5xx errors? I'm unsure as to the APi architecture generally and what the plans are for it as we develop CSIP.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppschallengesupportinterventionplanapi/client/manageusers/ManageUsersClient.kt [hunk 2]
```diff
@@ -16,6 +17,7 @@ class ManageUsersClient(@Qualifier("manageUsersWebClient") private val webClient
         .uri("/users/{username}", username)
         .retrieve()
         .bodyToMono(UserDetailsDto::class.java)
+        .retryNetworkExceptions()
```
Comment (ReedSoftware) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppschallengesupportinterventionplanapi/client/manageusers/ManageUsersClient.kt:20]: Is there some way we can add this as an extension to WebClient if we're wanting it generally? I'd just be worried someone will forget it

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppschallengesupportinterventionplanapi/client/RetryNetworkExceptions.kt (hunk 1 above)
Comment (iterative-it) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppschallengesupportinterventionplanapi/client/RetryNetworkExceptions.kt:9]: This is a fair concern. The retry mechanism is only used and indeed should only be used for idempotent, read only downstream endpoints. Getting user or prisoner information in this case. It's probably worth renaming this to make it's purpose clear. Something like `retryReadOnlyRequest` or even `retryIdempotentEvent` with some function documentation

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppschallengesupportinterventionplanapi/client/manageusers/ManageUsersClient.kt (hunk 2 above)
Comment (iterative-it) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppschallengesupportinterventionplanapi/client/manageusers/ManageUsersClient.kt:20]: Related to above comment, it does need to be an optional include cause we can't guarantee all downstream endpoints are idempotent

**Review by iterative-it — APPROVED** [review]
LGTM

**Pull request merged**